src/mon/NVMeofGwSerialize.h

```diff
@@ -0,0 +1,594 @@
+    }
+    return os;
+}
+
+inline std::ostream& print_gw_created_t(std::ostream& os, const NvmeGwCreated value, size_t num_ana_groups, NvmeAnaGrpId *anas) {
+    os << "==Internal map ==NvmeGwCreated { ana_group_id " << value.ana_grp_id << " osd_epochs: ";
+    for(size_t i = 0; i < num_ana_groups; i ++){
+        os << " " << anas[i] <<": " << value.blocklist_data[anas[i]].osd_epoch << ":" <<value.blocklist_data[anas[i]].is_failover ;
+    }
+    os << "\n" << MODULE_PREFFIX << "nonces: " << value.nonce_map << " }";
+    os << "\n" << MODULE_PREFFIX << "saved-nonces: " << value.copied_nonce_map << " }";
+    for (size_t i = 0; i < num_ana_groups; i++) {
+        os << " " << anas[i] <<": " << value.sm_state[anas[i]] << ",";
+    }
+    os <<  "]\n"<< MODULE_PREFFIX << " failover peers ";
+    for (size_t i = 0; i < num_ana_groups; i++) {
+        os << anas[i] <<": "  << value.failover_peer[anas[i]] << ",";
+    }
+    os << "]\n"<< MODULE_PREFFIX << "availability " << value.availability << "]";
+
+    return os;
+}
+
+inline std::ostream& operator<<(std::ostream& os, const NvmeGwCreated value) {
+    os << "==Internal map ==G W_CREATED_T { ana_group_id " << value.ana_grp_id << " osd_epochs: ";
+    for(int i = 0; i < MAX_SUPPORTED_ANA_GROUPS; i ++){
+        os << " " << value.blocklist_data[i].osd_epoch;
+    }
+    os << "\n" << MODULE_PREFFIX << "nonces: " << value.nonce_map << " }";
+
+    for (int i = 0; i < MAX_SUPPORTED_ANA_GROUPS; i++) {
+        os << value.sm_state[i] << ",";
+    }
+    os <<  "]\n"<< MODULE_PREFFIX << " failover peers ";
+    for (int i = 0; i < MAX_SUPPORTED_ANA_GROUPS; i++) {
+        os << value.failover_peer[i] << ",";
+    }
+    os <<  "]\n"<< MODULE_PREFFIX << " beacon-subsystems ";
+    for (const auto& sub: value.subsystems) {
+        os << sub << ",";
+    }
+
+    os << "]\n"<< MODULE_PREFFIX << "availability " << value.availability << "]";
+
+    return os;
+}
+
+inline std::ostream& operator<<(std::ostream& os, const NvmeGwCreatedMap value) {
+    if(value.size()) os << "\n" << MODULE_PREFFIX;;
+    NvmeAnaGrpId anas[MAX_SUPPORTED_ANA_GROUPS];
+    int i = 0;
+    for(auto &it: value ){
+       anas[i++] = it.second.ana_grp_id; // effective ana groups for these GWs within group_pool
+    }
+    for (auto &gw_created_map : value) {
+        os  <<  "gw_id: " << gw_created_map.first  << " [ " ;//<< gw_created_map.second << "] \n"<< MODULE_PREFFIX;
+        print_gw_created_t(os, gw_created_map.second,  value.size(), anas);
+        os << "] \n"<< MODULE_PREFFIX;
+    }
+    return os;
+}
+
+inline std::ostream& operator<<(std::ostream& os, const NVMeofGwMap value) {
+    os << "NVMeofGwMap [ Created_gws: ";
+    for (auto& group_gws: value.Created_gws) {
+        os <<  "\n" <<  MODULE_PREFFIX  << "{ " << group_gws.first << " } -> { " << group_gws.second << " }";
+    }
+    os << "]";
+    return os;
+}
+
+inline void encode(const ANA_STATE& st,  ceph::bufferlist &bl) {
```

Comment:
Replying to https://github.com/ceph/ceph/pull/54671#issuecomment-2159949922
@baum 
I'm not really sure what Rust's type encoding has to do here, but as I'm saying we have *lots* of helpers to make this all as seamless as possible. I don't think this function needs to exist at all, which is why I'm asking about it.

I asked (elsewhere) about separating the serialization because, as you've seen from this review process, it is both non-idiomatic in the Ceph codebase and very, very difficult to validate that the encodings are correct when they live so far from the type definitions.